packages/frontend/src/utils/major2-validation.ts

```diff
@@ -0,0 +1,672 @@
+    });
+  }
+
+  hasEnoughCoursesForBoth(s1: Solution, s2: Solution) {
+    let s1map = Major2ValidationTracker.createTakenMap(s1);
+    let s2map = Major2ValidationTracker.createTakenMap(s2);
+    // iterate through the solution with fewer courses for speed
+    let [fst, snd] =
+      s1.sol.length < s2.sol.length ? [s1map, s2map] : [s2map, s1map];
+    // for all courses in both solutions, check we have enough courses
+    for (let [cs, fstCount] of fst) {
+      let sndCount = snd.get(cs);
+      // if not in second solution, we have enough (skip)
+      if (!sndCount) continue;
+      const neededCount = fstCount + sndCount;
+      const tup = this.currentCourses.get(cs);
+      if (!tup) {
+        throw new Error("Solution contained a course that the tracker did not");
+      }
+      const actualCount = tup[1];
+      // if we don't have enough, return false, otherwise continue
+      if (actualCount < neededCount) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static createTakenMap(s: Solution): Map<string, number> {
+    const map = new Map();
+    for (const c of s.sol) {
+      const val = map.get(c) ?? 0;
+      map.set(c, val + 1);
+    }
+    return map;
+  }
+}
+
+// the result of major validation
+type MajorValidationResult = {
+  // true if no errors, false otherwise
+  ok: boolean;
+  majorRequirementsError: MajorValidationError | null;
+  totalCreditsRequirementError: TotalCreditsRequirementError | null;
+  // list of possible solutions, or null of no solution
+  solutions: Solution[] | null;
+};
+
+// concentrations are specified by their name or index in the concentrations list
+export function validateMajor2(
+  major: Major2,
+  taken: ScheduleCourse[],
+  concentrations?: string | number | (string | number)[]
+): MajorValidationResult {
+  const tracker = new Major2ValidationTracker(taken);
+  const concentrationReq = getConcentrationsRequirement(
+    concentrations,
+    major.concentrations
+  );
+  const majorReqs = [
+    ...major.requirementSections.map(sectionToReq2),
+    ...concentrationReq,
+  ];
+  // create a big AND requirement of all the sections and selected concentrations
+  const requirementsResult = validateRequirement(
+    {
+      type: "AND",
+      courses: majorReqs,
+    },
+    tracker
+  );
+  const creditsResult = validateTotalCreditsRequired(
+    major.totalCreditsRequired,
+    taken
+  );
+
+  let [solutions, majorRequirementsError] =
+    requirementsResult.type === "OK"
+      ? [requirementsResult.ok, null]
+      : [null, requirementsResult.err];
+  let totalCreditsRequirementError =
+    creditsResult.type === "OK" ? null : creditsResult.err;
+  return {
+    ok: !Boolean(totalCreditsRequirementError ?? majorRequirementsError),
+    solutions,
+    majorRequirementsError,
+    totalCreditsRequirementError,
+  };
+}
+
+export function getConcentrationsRequirement(
+  inputConcentrations: undefined | string | number | (string | number)[],
```

Comment:
what would the number type be used for? it seems like in that case it's checking against the index of the potential concentration(s) that could match the one(s) we're looking for but why would we need to check that way? is there a case where a concentration doesn't have a title?

Reply:
great question! there isn't actually a specific case.

i don't remember where i learned this, but in general, it is good API design to be flexible with inputs, and strict with outputs. this allows us to use this function in cases where we only know the index, as well as cases where we know the concentration name.

it isn't really necessary, but i thought it would be good to add since it's easy enough. what are your thoughts?

Reply:
cc @cindy1u0

Reply:
hmm i cant argue with good design practices hehe but on first glance i was pretty confused bc i couldn't (and sorta still can't) see a case where we would know the index/want to query by index.
buuut im not knowledgeable enough to have a strong opinion either way, so im good with it :D. if it's not a common best practices thing it might be better to have a tiny comment that explains why it's there to clarify any confusion in the future?